pkg/adsc/delta.go

```diff
@@ -0,0 +1,739 @@
+	if err := c.Dial(); err != nil {
+		return fmt.Errorf("dial context: %v", err)
+	}
+
+	xds := discovery.NewAggregatedDiscoveryServiceClient(c.conn)
+	xdsClient, err := xds.DeltaAggregatedResources(ctx, grpc.MaxCallRecvMsgSize(math.MaxInt32))
+	if err != nil {
+		return fmt.Errorf("delta stream: %v", err)
+	}
+	c.sendNodeMeta.Store(true)
+	c.xdsClient = xdsClient
+	go c.handleRecv()
+	for _, w := range c.initialWatches {
+		c.request(w)
+	}
+	return nil
+}
+
+func (c *Client) Dial() error {
+	conn, err := dialWithConfig(&c.cfg.Config)
+	if err != nil {
+		return err
+	}
+	c.conn = conn
+	return nil
+}
+
+// reconnect will create a new stream
+func (c *Client) reconnect() {
+	c.mutex.RLock()
+	if c.closed {
+		c.mutex.RUnlock()
+		return
+	}
+	c.mutex.RUnlock()
+
+	err := c.Run(context.Background())
+	if err != nil {
+		time.AfterFunc(c.cfg.BackoffPolicy.NextBackOff(), c.reconnect)
+	}
+}
+
+type Option func(c *Client)
+
+func NewDelta(discoveryAddr string, config *DeltaADSConfig, opts ...Option) *Client {
+	if config == nil {
+		config = &DeltaADSConfig{}
+	}
+	config.Address = discoveryAddr
+	config.Config = setDefaultConfig(&config.Config)
+	c := &Client{
+		cfg:             config,
+		handlers:        map[string]HandlerFunc{},
+		tree:            map[resourceKey]resourceNode{},
+		errChan:         make(chan error, 10),
+		deltaXDSUpdates: make(chan *discovery.DeltaDiscoveryResponse, 100),
+		received:        map[string]*discovery.DeltaDiscoveryResponse{},
+		mutex:           sync.RWMutex{},
+	}
+	for _, o := range opts {
+		o(c)
+	}
+	return c
+}
+
+func NewDeltaWithBackoffPolicy(discoveryAddr string, config *DeltaADSConfig, backoffPolicy backoff.BackOff, opts ...Option) *Client {
+	if config == nil {
+		config = &DeltaADSConfig{}
+	}
+	delta := NewDelta(discoveryAddr, config, opts...)
+	delta.cfg.BackoffPolicy = backoffPolicy
+	return delta
+}
+
+func typeName[T proto.Message]() string {
+	ft := new(T)
+	return resource.APITypePrefix + string((*ft).ProtoReflect().Descriptor().FullName())
+}
+
+// Register registers a handler for a type which is reflected by the proto message.
+func Register[T proto.Message](f func(ctx HandlerContext, res T, event Event)) Option {
+	return func(c *Client) {
+		c.handlers[typeName[T]()] = func(ctx HandlerContext, res proto.Message, event Event) {
+			f(ctx, res.(T), event)
+		}
+	}
+}
+
+// Watch registers an initial watch for a type based on the type reflected by the proto message.
+func Watch[T proto.Message](resourceName string) Option {
+	return initWatch(typeName[T](), resourceName)
+}
+
+func initWatch(typeURL string, resourceName string) Option {
+	return func(c *Client) {
+		if resourceName == "*" {
+			// Normalize to allow both forms
+			resourceName = ""
+		}
+		key := resourceKey{
+			Name:    resourceName,
+			TypeURL: typeURL,
+		}
+		existing, f := c.tree[key]
+		if f {
+			// We are watching directly now, so erase any parents
+			existing.Parents = nil
```

Comment:
should we erase child as well, initWatch is called after a new connect or reconnect. Seems a totally new initialization

Reply:
I think this is reasonable, the relationship will be rebuilt after processing responses.